File: mvvm/JetLiYu/app/src/main/java/pri/weiqiang/jet/liyujapanese/ui/adapter/WordAdapter.java
package pri.weiqiang.jet.liyujapanese.ui.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import pri.weiqiang.jet.liyujapanese.R;
import pri.weiqiang.jet.liyujapanese.data.bean.Word;


public class WordAdapter extends ListAdapter<Word, WordAdapter.ViewHolder> {

    private static final String TAG = WordAdapter.class.getSimpleName();
    // 列表展开标识
    int opened = -1;
    ArrayList<Integer> posOpened = new ArrayList<Integer>();
    OnItemClickListener onItemClickListener;
    Boolean isExpandable;

    public WordAdapter(Boolean isExpandable) {
        super(DIFF_CALLBACK);
        this.isExpandable = isExpandable;
    }

    private static final DiffUtil.ItemCallback<Word> DIFF_CALLBACK = new DiffUtil.ItemCallback<Word>() {
        @Override
        public boolean areItemsTheSame(@NonNull Word oldItem, @NonNull Word newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Word oldItem, @NonNull Word newItem) {
            return oldItem.getWord().equals(newItem.getWord());
        }
    };

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_word, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {

        holder.bean = getItem(position);
        holder.mTvWord.setText(holder.bean.getWord());
        holder.mTvPhonetic.setText(String.valueOf(holder.bean.getPhonetic()));
        holder.mTvTranslation.setText(holder.bean.getTranslation());
        holder.bind(position);

    }
    //使用RecyclerAdapter时使用
//    public void updateWordList(List<Word> list){
//        if(this.list!=null){
//            this.list.clear();
//            this.list.addAll(list);
//        }
//    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }


    public interface OnItemClickListener {

        void onItemClick(Word bean);

    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        public final View view;
        public final TextView mTvWord;
        public final TextView mTvPhonetic;
        public final TextView mTvTranslation;
        public final ImageView mIvMark;
        public final ImageView mIvTTS;

        public Word bean;

        public ViewHolder(View itemView) {
            super(itemView);

            this.view = itemView;
            this.mTvWord = itemView.findViewById(R.id.tv_word);
            this.mTvPhonetic = itemView.findViewById(R.id.tv_phonetic);
            this.mTvTranslation = itemView.findViewById(R.id.tv_translation);
            this.mIvMark = itemView.findViewById(R.id.iv_mark);
            this.mIvTTS = itemView.findViewById(R.id.iv_tts);
            itemView.setOnClickListener(this);
            mIvMark.setOnClickListener(this);
            mIvTTS.setOnClickListener(this);
        }

        // 此方法实现列表的展开和关闭
        public void bind(int pos) {
            if (bean.getFav() == 0) {
                this.mIvMark.setImageResource(R.drawable.ic_bookmark_border_primary);
            } else {
                this.mIvMark.setImageResource(R.drawable.ic_bookmark_primary);
            }
            if (isExpandable) {
                this.mTvWord.setText(bean.getWord());
                this.mTvPhonetic.setText(bean.getPhonetic());
                this.mTvPhonetic.setVisibility(View.VISIBLE);
                this.mTvTranslation.setVisibility(View.VISIBLE);
            } else if (posOpened.contains(pos)) {
                this.mTvWord.setText(bean.getWord());
                this.mTvPhonetic.setText(bean.getPhonetic());
                this.mTvTranslation.setVisibility(View.VISIBLE);
                this.mTvPhonetic.setVisibility(View.VISIBLE);
            } else {
                this.mTvWord.setText(bean.getWord());
                this.mTvPhonetic.setVisibility(View.GONE);
                this.mTvTranslation.setVisibility(View.GONE);
            }
        }

        /**
         * recyclerView是不提供onItemClickListener的。所以列表的点击事件需要我们自己来实现
         *
         * @param v
         */
        @Override
        public void onClick(View v) {
            Log.e(TAG, "getAdapterPosition:" + getAdapterPosition() + ",opened:" + opened);
            switch (v.getId()) {
                case R.id.iv_mark:
                    Log.e(TAG, "iv_mark onClick.");
//                    DBManager.getInstance().updateFav(bean);
                    if (bean.getFav() == 0) {
                        bean.setFav(1);
//                        DBManager.getInstance().insertFav(bean);
                    } else {
                        bean.setFav(0);
//                        DBManager.getInstance().delFav(bean);
                    }
                    break;
                case R.id.iv_tts:
                    Log.e(TAG, "iv_tts onClick.");
                    break;
                default:
                    Log.e(TAG, "itemView onClick.");
                    if (posOpened.contains(getAdapterPosition())) {
                        posOpened.remove(posOpened.indexOf(getAdapterPosition()));
                    } else {
                        posOpened.add(getAdapterPosition());
                    }
            }
            notifyItemChanged(getAdapterPosition());

        }
    }




}

File: mvvm/KtJetLiYu/app/src/main/java/pri/weiqiang/jet/kt/liyujapanese/ui/adapter/WordAdapter.kt
package pri.weiqiang.jet.kt.liyujapanese.ui.adapter

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import pri.weiqiang.jet.kt.liyujapanese.R
import pri.weiqiang.jet.kt.liyujapanese.data.bean.Word

class WordAdapter(private var isExpandable: Boolean) :
    ListAdapter<Word, WordAdapter.ViewHolder>(DiffCallback) {

    //    var posOpened:ArrayList<Int> = ArrayList<Int>()
//    var posOpened:ArrayList<Int> = ArrayList()
    var posOpened = ArrayList<Int>()
    var opened = -1

    companion object {
        val TAG = WordAdapter::class.java.simpleName
    }

    object DiffCallback : DiffUtil.ItemCallback<Word>() {
        override fun areItemsTheSame(oldItem: Word, newItem: Word): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Word, newItem: Word): Boolean {
            return oldItem.word == newItem.word
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view: View =
            LayoutInflater.from(parent.context).inflate(R.layout.item_word, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bean = getItem(position)
        holder.mTvWord.setText(holder.bean?.word)
        holder.mTvPhonetic.setText(holder.bean?.phonetic)
        holder.mTvTranslation.setText(holder.bean?.translation)
        holder.bind(position)
    }

    inner class ViewHolder(val itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        val mTvWord: TextView
        val mTvPhonetic: TextView
        val mTvTranslation: TextView
        val mIvMark: ImageView
        val mIvTTS: ImageView
        var bean: Word? = null

        override fun onClick(v: View?) {
            Log.e(TAG, "getAdapterPosition:" + getAdapterPosition() + ",opened:" + opened)
            when (v?.id) {
                R.id.iv_mark -> {
                    Log.e(TAG, "iv_mark onClick.")
                    //                    DBManager.getInstance().updateFav(bean);
                    if (bean?.fav == 0) {
                        bean?.fav = 1
//                        DBManager.getInstance().insertFav(bean);
                    } else {
                        bean?.fav = 0
//                        DBManager.getInstance().delFav(bean);
                    }
                }
                R.id.iv_tts -> {
                    Log.e(TAG, "iv_tts onClick.")
                }
                else -> {
                    Log.e(TAG, "itemView onClick.")
                    if (posOpened.contains(adapterPosition)) {
                        posOpened.remove(posOpened.indexOf(adapterPosition))
                    } else {
                        posOpened.add(adapterPosition)
                    }
                }


            }
            notifyItemChanged(adapterPosition)
        }

        init {
            mTvWord = itemView.findViewById(R.id.tv_word)
            mTvPhonetic = itemView.findViewById(R.id.tv_phonetic)
            mTvTranslation = itemView.findViewById(R.id.tv_translation)
            mIvMark = itemView.findViewById(R.id.iv_mark)
            mIvTTS = itemView.findViewById(R.id.iv_tts)
            itemView.setOnClickListener(this)
            mIvMark.setOnClickListener(this)
            mIvTTS.setOnClickListener(this)
        }

        fun bind(pos: Int) {
            if (bean?.fav == 0) {
                this.mIvMark.setImageResource(R.drawable.ic_bookmark_border_primary)
            } else {
                this.mIvMark.setImageResource(R.drawable.ic_bookmark_primary)
            }
            if (isExpandable) {
                this.mTvWord.setText(bean?.word)
                this.mTvPhonetic.setText(bean?.phonetic)
                this.mTvPhonetic.setVisibility(View.VISIBLE)
                this.mTvTranslation.setVisibility(View.VISIBLE)
            } else if (posOpened.contains(pos)) {
                this.mTvWord.setText(bean?.word)
                this.mTvPhonetic.setText(bean?.phonetic)
                this.mTvTranslation.setVisibility(View.VISIBLE)
                this.mTvPhonetic.setVisibility(View.VISIBLE)
            } else {
                this.mTvWord.setText(bean?.word)
                this.mTvPhonetic.setVisibility(View.GONE)
                this.mTvTranslation.setVisibility(View.GONE)
            }
        }
    }
}